### PR TITLE
stabilize ptr_offset_from

### LINKED ISSUE #41079: Tracking issue for ptr::offset_to
PR: #40943

Adds an `offset_to` method to calculate the distance between two raw pointers.

### PR DESCRIPTION
**Pull request by RalfJung**
This stabilizes ptr::offset_from, and closes https://github.com/rust-lang/rust/issues/41079. It also removes the deprecated `wrapping_offset_from`. This function was deprecated 19 days ago and was never stable; given an FCP of 10 days and some waiting time until FCP starts, that leaves at least a month between deprecation and removal which I think is fine for a nightly-only API.

Regarding the open questions in https://github.com/rust-lang/rust/issues/41079:
* Should offset_from abort instead of panic on ZSTs? -- As far as I know, there is no precedent for such aborts. We could, however, declare this UB. Given that the size is always known statically and the check thus rather cheap, UB seems excessive.
* Should there be more methods like this with different restrictions (to allow nuw/nsw, perhaps) or that return usize (like how isize-taking offset is more conveniently done with usize-taking add these days)? -- No reason to block stabilization on that, we can always add such methods later.

Also nominating the lang team because this exposes an intrinsic.

The stabilized method is best described [by its doc-comment](https://github.com/RalfJung/rust/blob/56d4b2d69abb93e4f0ca79471deca7aaaaeca214/src/libcore/ptr/const_ptr.rs#L227). The documentation forgot to mention the requirement that both pointers must "have the same provenance", aka "be derived from pointers to the same allocation", which I am adding in this PR. This is a precondition that [Miri already implements](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=a3b9d0a07a01321f5202cd99e9613480) and that, should LLVM ever obtain a `psub` operation to subtract pointers, will likely be required for that operation (following the semantics in [this paper](https://people.mpi-sws.org/~jung/twinsem/twinsem.pdf)).

### CODE REVIEW

File: src/libcore/ptr/const_ptr.rs [hunk 1]
```diff
@@ -286,7 +289,26 @@ impl<T: ?Sized> *const T {
     ///     assert_eq!(ptr2.offset(-2), ptr1);
     /// }
     /// ```
-    #[unstable(feature = "ptr_offset_from", issue = "41079")]
+    ///
+    /// *Incorrect* usage:
+    ///
+    /// ```rust,no_run
+    /// #![feature(ptr_offset_from)]
```
Comment (lzutao) [inline src/libcore/ptr/const_ptr.rs:296]: If this method is stabilized, this attr is no need.

**Comment by Amanieu** [comment]
@rfcbot fcp merge

**Comment by scottmcm** [comment]
> Also nominating the lang team because this exposes an intrinsic.

This used to just be implemented with normal code, IIRC.  Is there some observable semantics that the intrinsic provides?  Or is it just for optimization/miri convenience?

**Comment by RalfJung** [comment]
Hm... so in terms of const-eval, there is definitely something observable, as the implementation with normal code just does not work. (CTFE stops when you try to subtract integers that were obtained from pointers, as this is not an operation we can support in general.)

But you are right that the part that is being stabilized here could actually be implemented (possibly less efficiently) without the intrinsic. I had not thought of that.

**Comment by oli-obk** [comment]
>     * Should offset_from abort instead of panic on ZSTs? -- As far as I know, there is no precedent for such aborts. We could, however, declare this UB. Given that the size is always known statically and the check thus rather cheap, UB seems excessive.

There's no precedent for this, but we also have a third option: make it a post-monomorphization error. Doing this will cause very suboptimial diagnostics at the moment, but it's certainly an option (also I've started looking into finding a reasonable way to make these diagnostics nicer since then).

```rust
#![feature(const_type_name)]
#![feature(const_transmute)]
#![feature(const_panic)]

const IS_ZST: &[u8] = b" is zst";

trait ZstAssert: Sized {
    #[warn(const_err)]
    const CHECK: () = if std::mem::size_of::<Self>() == 0 {
        // would like to use the length of `IS_ZST` + the length of the
        // return value of `type_name::<Self>()` here, but that won't work because you can't use
        // generic parameters in array lengths, not even in repeat expression lengths.
        let mut y = [0_u8; 1000];
        let name: &[u8] = std::any::type_name::<Self>().as_bytes();
        let mut i = 0;
        while i < name.len() {
            y[i] = name[i];
            i += 1;
        }
        while i < IS_ZST.len() + name.len() {
            y[i] = IS_ZST[i - name.len()];
            i += 1;
        }

        let x: &str = unsafe { std::mem::transmute(&y as &[u8]) };
        panic!(x)
    };
}

impl<T: Sized> ZstAssert for T {}

fn foo<T: Sized>() {
    #[allow(path_statements)]
    {
        T::CHECK;
    }
}

fn main() {
    foo::<i32>();
    foo::<()>();
}
```

gives us 

```
warning: any use of this value will cause an error
  --> src/main.rs:26:9
   |
9  | /     const CHECK: () = if std::mem::size_of::<Self>() == 0 {
10 | |         // would like to use the length of `IS_ZST` + the length of the
11 | |         // return value of `type_name::<Self>()` here, but that won't work because you can't use
12 | |         // generic parameters in array lengths, not even in repeat expression lengths.
...  |
26 | |         panic!(x)
   | |         ^^^^^^^^^ the evaluated program panicked at '() is zst', src/main.rs:26:9
27 | |     };
   | |______-
   |
note: the lint level is defined here
  --> src/main.rs:8:12
   |
8  |     #[warn(const_err)]
   |            ^^^^^^^^^
   = note: this warning originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: erroneous constant encountered
  --> src/main.rs:35:9
   |
35 |         T::CHECK;
   |         ^^^^^^^^

error: aborting due to previous error; 1 warning emitted
```

which doesn't show us the call site of `foo` (which is what I mean by the suboptimal diagnostic).

**Comment by RalfJung** [comment]
As you noted in Zulip, the problem with this is that it breaks code that specifically does `if size:of::<T>() > 0` first.

**Comment by oli-obk** [comment]
Well, it doesn't *break* code, since `ptr_offset_from` is not stable yet. It will start requiring any users of `ptr_offset_from` to use similar trait based schemes to differentiate between the zst logic and non-zst logic. I don't think this is feasible right now, as I can't come up with a way to use the same thing for changing some runtime code that will get executed *without* also triggering the wrong monomorphization. So marking my alternative as resolved.

**Comment by nikomatsakis** [comment]
Removing nomination since we discussed in today's @rust-lang/lang meeting. No concerns were raised.

**Comment by nikomatsakis** [comment]
cc @cramertj @pnkfelix @withoutboats @sfackler for remaining checkboxes :)

**Comment by nikomatsakis** [comment]
cc @pnkfelix @cramertj and @sfackler for check-boxes on this issue

**Comment by RalfJung** [comment]
@pnkfelix @cramertj  @sfackler any chance you could take a look at this?

Is there any kind of regular schedule or so on which we could put this to get people to look at it?

**Review by cuviper — COMMENTED** [review]
The stable versions should be 1.47, unless you're planning a really quick beta backport...

**Comment by scottmcm** [comment]
@RalfJung No, I don't think I can come up with a reason.  And I guess with `ptr::raw!` and friends coming along unaligned-but-separated pointers are plausible.

All good as-is, then.

**Pull request merged**